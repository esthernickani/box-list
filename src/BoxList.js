import React, { useState } from "react";
import { v4 as uuid } from 'uuid';
import Box from './Box'
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
    const [boxes, setBoxes] = useState([]);
    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, { ...newBox, id: uuid() }])
    }


    const handleRemove= (id) => {
        const newBoxList = boxes.filter(box => box.id !== id)
        setBoxes(newBoxList)
    }

    const boxComponents = boxes.map(({ id, width, height, backgroundColor }) => <Box width={width} height={height} backgroundColor={backgroundColor} key={id} id={id} handleRemove={handleRemove} />)

    return (
        <div>
            <h1>Boxes</h1>
            <NewBoxForm addBox={addBox} />
            {boxComponents}
        </div>
    )
}

export default BoxList;