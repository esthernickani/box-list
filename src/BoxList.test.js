import React from "react";
import BoxList from "./BoxList";
import { fireEvent, render } from "@testing-library/react";

const addBox = (boxList, height = 15, width = 15, backgroundColor = "olivegreen") => {
    const widthInput = boxList.getByLabelText("Width")
    const heightInput = boxList.getByLabelText("height")
    const backgroundColorInput = boxList.getByLabelText("Background Color")
    const btn = boxList.getByText("Submit")

    fireEvent.change(widthInput, { target: { value: width}})
    fireEvent.change(heightInput, { target: { value: height}})
    fireEvent.change(backgroundColorInput, { target: { value: backgroundColor}})

    fireEvent.click(btn);
}

it("renders without crashing", function() {
    render(<BoxList />)
})

it("matches snapshot", function() {
    const { asFragment } = render(<BoxList />)
    expect(asFragment()).toMatchSnapshot()
})

it("adds new box", function() {
    const boxList = render(<BoxList />)

    const removeBtn = boxList.queryByText("remove")
    expect(removeBtn).not.toBeInTheDocument();

    addBox(boxList)

    const remove = boxList.getByText("remove")
    expect(remove).toBeInTheDocument();
    expect(remove.previousSibling).toHaveStyle(`
        width: 15px;
        height: 15px;
        background-color: olivegreen;
    `)

})

it("removes a box", function() {
    const boxList = render(<BoxList />)
    addBox(boxList)

    const removeBtn = boxList.getByText("remove")
    fireEvent.click(removeBtn)

    expect(removeBtn).not.toBeInTheDocument()
})