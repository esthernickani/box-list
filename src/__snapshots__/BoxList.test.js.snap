// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Boxes
    </h1>
    <form>
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        placeholder="Width"
        type="number"
        value=""
      />
      <label
        for="height"
      >
        height
      </label>
      <input
        id="height"
        name="height"
        placeholder="height"
        type="number"
        value=""
      />
      <label
        for="backgroundColor"
      >
        Background Color
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        placeholder="Background Color"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
